KAFKA-6361: Fix log divergence between leader and follower after fast leader fail over

apovzner opened the pull request:
Implementation of KIP-279 as described here: https://cwiki.apache.org/confluence/display/KAFKA/KIP-279%3A+Fix+log+divergence+between+leader+and+follower+after+fast+leader+fail+over

In summary:
- Added leader_epoch to OFFSET_FOR_LEADER_EPOCH_RESPONSE
- Leader replies with the pair( largest epoch less than or equal to the requested epoch, the end offset of this epoch)
- If Follower does not know about the leader epoch that leader replies with, it truncates to the end offset of largest leader epoch less than leader epoch that leader replied with, and sends another OffsetForLeaderEpoch request. That request contains the largest leader epoch less than leader epoch that leader replied with.

Added integration test EpochDrivenReplicationProtocolAcceptanceTest.logsShouldNotDivergeOnUncleanLeaderElections that does 3 fast leader changes where unclean leader election is enabled and min isr is 1. The test failed before the fix was implemented.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apovzner committed:
KAFKA-6361: Fix log divergence between leader and follower after fast leader fail over

apovzner committed:
set EpochEndOffset.UNDEFINED_EPOCH default to NO_PARTITION_LEADER_EPOCH (same -1 as before)

apovzner committed:
some cleanup and tests

apovzner committed:
Few small fixes to truncation logic

apovzner committed:
fixed merge errors

apovzner commented:
@lindong28 and @junrao Regarding the truncation logic for future replica (ReplicaAlterLogDirsThread): I thought about this a bit more, and I think we don't need the same truncation logic as in the replica fetcher. We can implement a much simpler logic based on the truncation offset of the local replica. Here are main points:
1) Currently (and in the original implementation), if the replica becomes a follower, it does a truncation and forces the truncation state on the future replica and provides it with its truncation offset. The original implementation and this PR right now does not use that offset except if the future or current replica does not have any epochs recorded. 
2) I think it is safe for the future replica to truncate and fetch from the offset that the local replica truncated to, i.e., min of that offset and future replica's LOE. The reason is that the future replica only ever fetches from the local replica == one source. So, the only reason why future replica and local replica's logs may diverge is if the local replica truncates and fetches different offsets from the new leader. In this case, we already force truncation of the future replica and provide the new truncate offset of the local replica.
3) One possible reason against is that what if we "lose" future epoch sequence file or entries from it. The leader epoch truncation logic may help if we lose some recent entries because it will truncate to the end offset of the last known epoch. However, it does not guarantee proper recovery in many cases. Plus, I understand that the future replica will do same recovery logic as normal replica if say something gets corrupted, or we lose the leader epoch file, e.g., someone deletes it (it gets rebuilt from the log). 

Based on the above, unless I missed something or my assumptions are wrong, I think we should truncate future replica to the truncation offset of the local replica. This will result in a simpler code/logic which is easier to reason about and debug in the future.

The PR is also ready to review.

lindong28 commented:
@apovzner Thanks for the patch! I will finish the first round of review this week.

lindong28 commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +276,47 @@ abstract class AbstractFetcherThread(name: String,
+          else
+            "follower"
+        info(
+          s"Based on $followerName's leader epoch, leader replied with an offset $logEndOffset >= the " +
```
nits: `leader replied with an offset $offsetToTruncateTo` not (logEndOffset)

apovzner replied:
Oh that's a good catch, and also made me realize it's not a completely correct log message for all the cases, will fix.

lindong28 commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -290,9 +291,12 @@ class ReplicaFetcherThread(name: String,
     * - If the leader's offset is greater, we stick with the Log End Offset
     *   otherwise we truncate to the leaders offset.
     * - If the leader replied with undefined epoch offset we must use the high watermark
+    * - If the leader replied with leader epoch not known to this follower, we truncate to the
```
Since we are changing this Java doc, can we make this Java doc a bit more consistent with the actual implementation. For example, `If the leader replied with undefined epoch` can probably be the first case.

lindong28 commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -304,19 +308,50 @@ class ReplicaFetcherThread(name: String,
-          partition.truncateTo(fetchOffset, isFuture = false)
-          replicaMgr.replicaAlterLogDirsManager.markPartitionsForTruncation(brokerConfig.brokerId, tp, fetchOffset)
-          fetchOffsets.put(tp, fetchOffset)
+              OffsetTruncationState(partitionStates.stateValue(tp).fetchOffset, truncationCompleted = true)
```
nits: This is not related to this patch. Can we add one comment saying that the initial offset in this case will be the high watermark, so that it is more consistent with the Java doc?

apovzner replied:
Added comment above the logging

lindong28 commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -304,19 +308,50 @@ class ReplicaFetcherThread(name: String,
+              val (replicaLeaderEpoch, replicaEndOffset) = replica.epochs.get.endOffsetFor (epochOffset.leaderEpoch)
+              if (replicaEndOffset == UNDEFINED_EPOCH_OFFSET) {
+                // This can happen if replica was not tracking offsets at that point (before the
+                // upgrade, or if this broker is new). Since the leader replied with epoch <=
```
nits: Would it be a bit more accurate a bit more accurate to replace `<=` with `<`?

lindong28 commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -304,19 +308,50 @@ class ReplicaFetcherThread(name: String,
-          replicaMgr.replicaAlterLogDirsManager.markPartitionsForTruncation(brokerConfig.brokerId, tp, fetchOffset)
-          fetchOffsets.put(tp, fetchOffset)
+              OffsetTruncationState(partitionStates.stateValue(tp).fetchOffset, truncationCompleted = true)
+            } else if (epochOffset.leaderEpoch == UNDEFINED_EPOCH) {
```
nits: It probably does not affect the correctness of the code. But I am wondering if we can make `ReplicaFetcherThread.maybeTruncate()` a bit more consistent with the `ReplicaAlterDirThread.maybeTruncate()`. Currently `ReplicaFetcherThread.maybeTruncate()` will specifically handle the scenario that `epochOffset.leaderEpoch == UNDEFINED_EPOCH` whereas `ReplicaAlterDirThread.maybeTruncate()` handles the scenario through `futureEndOffset == UNDEFINED_EPOCH_OFFSET`.

Also, should we also consider to `min` with  the initial offset here just like this patch does in `ReplicaAlterDirThread.maybeTruncate()`? I am wondering whether we can make them more consistent, or whether there is reason that the initial offset is needed in only one of them.

apovzner replied:
About the question in the second paragraph, it would be incorrect to do min with the initial offset (high watermark) here, because this will fallback to pre-KIP-101 implementation and we can actually lose a committed message (see scenario 1 in KIP-101) . This particular case can happen if the leader is on the protocol version of pre this KIP but post-KIP-101, so it replies with the valid offset but invalid leader epoch. In this case, we want to do KIP-101 implementation of truncating to leader's offset, rather than falling back to pre-KIP-101 implementation. 

Regarding a bigger question of making `AlterLogDirThread` more consistent with `ReplicaAlterLogDirThread`, I wanted to discuss the possibility of `ReplicaAlterLogDirThread` using only initial offset (which is truncation offset of the main replica) for truncation instead of following offset for leader epoch logic. I left the comment earlier in this PR and wanted to get your opinion. 
-- The initial offset in `AlterLogDirThread` is main replica truncation offset (if main replica is a follower) or main replica's high watermark (if main replica is the leader), which is different from initial offset in `ReplicaFetcherThread` which is a high watermark.
-- There is no way that future replica needs to truncate further back than initial offset, because it is always a follower of the main replica, and if the main replica truncated and re-fetched offsets from the leader causing temporary log divergence with the future replica, we already force truncation on the future replica setting future replica's initial offset to the main replica truncation offset. 

Otherwise, I will change ReplicaAlterDirThread.maybeTruncate() to be more consistent with ReplicaAlterDirThread.maybeTruncate() in the case mentioned above. Although the current implementation is correct, because falling back to the initial offset for the future replica is safe (vs. for follower replica), because the future replica is always a follower. The whole reason for KIP-101 and KIP-279 is to deal with replicas changing their leader/follower status, while the future replica is always a follower.

lindong28 replied:
@apovzner Thanks for the explanation. The case that the leader is on the protocol version of pre this KIP but post-KIP-101 AND this patch is used in some broker, can only happen if the Kafka cluster being upgraded to use this patch. The time window of this state is very small and maybe we do not need to take care of this scenario.

Regarding the possibility of ReplicaAlterLogDirThread using only initial offset, Jun has provided a very good example. Basically this approach is not reliable if the future replica is offline when leader replica is truncated.

apovzner replied:
Actually, to be more precise, both leader and follower could be on pre- this KIP protocol version, if the user upgrades the brokers but do not bump the protocol version. So I think we want post KIP-101 behavior, which is what's implemented, vs. going back to pre-KIP-101.

OK, offline future replica is a good example I did not consider. I agree we should use the same algorithm in ReplicaAlterLogDirThread. I will make it more consistent.

lindong28 reviewed (commented):
Thanks for the patch! LGTM. I only have some minor comments regarding the Java doc and the consistency between ReplicaFetcherThread and ReplicaAlterDirThread.

apovzner committed:
Fixed build failures and fixed logging

apovzner committed:
Updated comments based on review comments

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequest.java:
```diff
@@ -50,8 +50,11 @@
+
     public static Schema[] schemaVersions() {
-        return new Schema[]{OFFSET_FOR_LEADER_EPOCH_REQUEST_V0};
+        return new Schema[]{OFFSET_FOR_LEADER_EPOCH_REQUEST_V0, OFFSET_FOR_LEADER_EPOCH_REQUEST_V1};
```
Since this affects inter broker protocol, we need to (1) document this api change for "2.0-IV0" in ApiVersion.scala, (2) update the upgrade section in the doc, (3) only use the new protocol if the inter broker protocol is 2.0-IV0 or above.

apovzner replied:
Done all three.

apovzner replied:
Regarding (3), the fetcher falls back to KIP-101 logic if inter-broker protocol version < KAFKA_2_0_IV0 (ignores leader epoch returned in the response and uses end offset).

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -286,13 +287,23 @@ class ReplicaFetcherThread(name: String,
-  override def maybeTruncate(fetchedEpochs: Map[TopicPartition, EpochEndOffset]): ResultWithPartitions[Map[TopicPartition, Long]] = {
-    val fetchOffsets = scala.collection.mutable.HashMap.empty[TopicPartition, Long]
+   * Truncate the log for each partition's epoch based on leader's returned epoch and offset.
+   *  -- If the leader replied with undefined epoch offset, we must use the watermark. This can
```
watermark => high watermark

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -286,13 +287,23 @@ class ReplicaFetcherThread(name: String,
-    val fetchOffsets = scala.collection.mutable.HashMap.empty[TopicPartition, Long]
+   * Truncate the log for each partition's epoch based on leader's returned epoch and offset.
+   *  -- If the leader replied with undefined epoch offset, we must use the watermark. This can
+   *  happen if 1) the leader is on the protocol version < KAFKA_0_11_0_IV2; 2) the follower
```
To be more precise, 1) should be "the leader is still using message format older than KAFKA_0_11_0_IV2".

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -286,13 +287,23 @@ class ReplicaFetcherThread(name: String,
+   *  requested leader epoch < the first leader epoch known to the leader.
+   *  -- If the leader replied with the valid offset but undefined leader epoch, we truncate to
+   *  leader's offset if it is lower than follower's Log End Offset. This may happen if the
+   *  leader in on the protocol version < KAFKA_1_1_IV0
```
KAFKA_1_1_IV0 should be  KAFKA_2_0_IV0

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -286,13 +287,23 @@ class ReplicaFetcherThread(name: String,
+   *  end offset of the largest epoch that is smaller than the epoch the leader replied with, and
+   *  send offset for leader epoch request with that leader epoch. In a more rare case, where the
+   *  follower was not tracking epochs smaller than the epoch the leader replied with, we
+   *  truncate the leader's offset (and do not send any more leader epoch requesrs).
```
I think in this case, it's probably better to fall back to high watermark. That way, if the leader epoch logic doesn't apply, we always consistently fall back to the old method.

apovzner replied:
My concern about falling back to high watermark in this particular case is that post-KIP-101 code (and pre-2.0) behaves exactly as described, since the leader does not send leader epoch, so we don't check it and use leader's offset to truncate. And also if brokers upgrade to 2.0, but do not upgrade protocol version. Then, we upgrade to 2.0 protocol version, and we are back to high watermark in this case.

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -355,26 +398,19 @@ class ReplicaFetcherThread(name: String,
           // if we get any unexpected exception, mark all partitions with an error
           result = partitions.map { case (tp, _) =>
-            tp -> new EpochEndOffset(Errors.forException(t), UNDEFINED_EPOCH_OFFSET)
+            tp -> new EpochEndOffset(Errors.forException(t), UNDEFINED_EPOCH, UNDEFINED_EPOCH_OFFSET)
```
Now that we have 2 different versions of the LeaderEpochRequest, we need to use the right version of the request depending on inter broker protocol version.

apovzner replied:
I have a question regarding this. In OffsetsForLeaderEpochResponse, I used `partitionStruct.setIfExists` and `partitionAndEpoch.getOrElse` for new leader epoch field vs. explicitly checking version for setting/getting fields. I think, it should be correct to e.g., return the leader epoch but make sure to use the older protocol in fetcher if the inter-broker protocol is below 2.0. Or is it preferred to check the version and not set the field in OffsetsForLeaderEpochResponse? I looked at the code, and it looks like different request/response use different approaches (either one I described).

This particular line of code is correct independent of version, since undefined offsets means falling back to high watermark, independent if epoch is set or not (or defined or not defined). 

I agree I need to update the truncation logic to make sure the behavior is consistent with the protocol version.

junrao replied:
Using setIfExists seems fine. However, we need to pass in the version number to OffsetsForLeaderEpochRequest.Builder based on inter broker protocol version. Similar to what we do when building the fetch request.

apovzner replied:
I looked at OffsetsForLeaderEpochRequest.Builder, and I see that we pass version to the builder:
`public OffsetsForLeaderEpochRequest build(short version) {`
` public OffsetsForLeaderEpochRequest(Struct struct, short version) {`
` @Override
    protected Struct toStruct() {
        Struct requestStruct = new Struct(ApiKeys.OFFSET_FOR_LEADER_EPOCH.requestSchema(version()));
... `
Or am I missing something?

apovzner replied:
Oh I see now, version is not in OffsetForLeaderEpochResponse constructor, not request. Will fix.

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -304,19 +315,51 @@ class ReplicaFetcherThread(name: String,
+              // this may happen if the leader used version 0 of OffsetForLeaderEpoch
+              // request/response
+              warn(s"Based on follower's leader epoch, leader replied with an unknown leader epoch in ${replica.topicPartition}. " +
+                   s"The leader's offset only ${epochOffset.endOffset} will be used for truncation.")
```
Should we remove "only"?

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -304,19 +315,51 @@ class ReplicaFetcherThread(name: String,
+            } else {
+              // get (leader epoch, end offset) pair that corresponds to the largest leader epoch
+              // less than or equal to the requested epoch.
+              val (replicaLeaderEpoch, replicaEndOffset) = replica.epochs.get.endOffsetFor (epochOffset.leaderEpoch)
```
Could we remove the space after endOffsetFor?

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -304,19 +315,51 @@ class ReplicaFetcherThread(name: String,
+                     s"Will truncate to $intermediateOffsetToTruncateTo and send another leader epoch request to the leader.")
+                OffsetTruncationState(intermediateOffsetToTruncateTo, truncationCompleted = false)
+              } else {
+                val offsetToTruncateTo = min(replicaEndOffset, epochOffset.endOffset)
```
Perhaps it's clearer if we rename epochOffset to leaderEpochOffset.

junrao commented on core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala:
```diff
@@ -81,33 +81,44 @@ class LeaderEpochFileCache(topicPartition: TopicPartition, leo: () => LogOffsetM
+        if (requestedEpoch == UNDEFINED_EPOCH) {
+          // this is ok in case of bootstrapping follower sending a request with undefined epoch
+          // since it does not have any epochs; but otherwise not expected. So, we log as info
+          info(s"Requested end offset for undefined leader epoch for partition $topicPartition. " +
```
Do we need to log this?

apovzner replied:
You are right, probably not here. I was thinking to log the case where the follower somehow loses the leader epoch sequence file or something that causes it sending the request with undefined epoch, which could potentially cause log divergence. But now I thinking we should log error on the follower if we detect such problem.

junrao commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -21,7 +21,7 @@ import java.nio.ByteBuffer
 
 import AbstractFetcherThread.ResultWithPartitions
-import kafka.cluster.BrokerEndPoint
+import kafka.cluster.{Replica, BrokerEndPoint}
```
unused import Replica.

junrao commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -186,16 +187,36 @@ class ReplicaAlterLogDirsThread(name: String,
+                val offsetToTruncateTo = min(futureEndOffset, epochOffset.endOffset)
+                finalFetchLeaderEpochOffset(offsetToTruncateTo, epochOffset.endOffset, futureReplica, isFutureReplica = true)
+              }
+            }
```
The above logic may still be needed in the following sequence: (1) future replica copies data above HW from current replica; (2) future replica is offline (e.g. disk failure); (3) current replica truncates data above HW and re-replicated new data from leader on the truncated offsets. To avoid duplicates, perhaps we can share the code between ReplicaFetchThread and here?

apovzner replied:
That's a good example I did not consider. In that case, I agree, we need the leader epoch logic. I will try to move this code out into a common method that both fetchers re-use.

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -647,15 +647,18 @@ class Partition(val topic: String,
   /**
     * @param leaderEpoch Requested leader epoch
-    * @return The last offset of messages published under this leader epoch.
+    * @return The last offset of messages published under this leader epoch, or if the the
```
typo "the the". Also, we are now returning the leader epoch and the end offset.

junrao commented on core/src/test/scala/unit/kafka/server/epoch/util/ReplicaFetcherMockBlockingSend.scala:
```diff
@@ -28,17 +28,27 @@ import org.apache.kafka.common.utils.{SystemTime, Time}
   var fetchCount = 0
   var epochFetchCount = 0
   var callback: Option[() => Unit] = None
+  var updatedOffsetsOpt: Option[java.util.Map[TopicPartition, EpochEndOffset]] = None
```
Would it be simpler to just initialize updatedOffsetsOpt to offsets and make it a none Option?

junrao commented on core/src/test/scala/unit/kafka/server/epoch/EpochDrivenReplicationProtocolAcceptanceTest.scala:
```diff
@@ -298,6 +299,86 @@ class EpochDrivenReplicationProtocolAcceptanceTest extends ZooKeeperTestHarness
+    brokers(0).startup()
+
+    //Bounce the producer (this is required, although I'm unsure as to why?)
+    producer.close()
```
This is probably because the broker port changes on restart?

junrao commented on core/src/test/scala/unit/kafka/server/epoch/LeaderEpochFileCacheTest.scala:
```diff
@@ -633,7 +644,7 @@ class LeaderEpochFileCacheTest {
 
     //Then
-    assertEquals(-1, cache.endOffsetFor(7))
+    assertEquals((UNDEFINED_EPOCH, -1), cache.endOffsetFor(7))
```
Change -1 to UNDEFINED_EPOCH_OFFSET?

junrao reviewed (commented):
@apovzner : Thanks for the patch. Left a few comments below.

apovzner committed:
Updated the upgrade doc and minor fixes in comments

apovzner committed:
Addressed review comments in tests

apovzner committed:
Truncate based on protocol version

apovzner committed:
Moved common truncation logic to method in base class

apovzner commented:
@junrao and @lindong28 Thanks a lot for your comments. I addressed all of them. 

Based on the use case of future replica being offline and missing "mark for truncation" event, I agree that `ReplicaAlterLogDirsThread` should use the same leader epoch logic for truncation as in `ReplicaFetherThread`. I moved the common logic that finds truncation offset to `AbstractFetcherThread.getOffsetTruncationState`, so now the truncation logic in both fetchers is consistent.

apovzner committed:
Added test for truncation logic for protocol version < 2.0

apovzner committed:
Added test for truncating to largest common epoch in ReplicaAlterLogDirsThread

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,88 @@ abstract class AbstractFetcherThread(name: String,
+
+    if (leaderEpochOffset.endOffset == UNDEFINED_EPOCH_OFFSET) {
+      // truncate to initial offset which is the high watermark for follower replica or current
+      // replica's truncation offset for the future replica (when the current replica truncates,
```
Hmm, in ReplicaManager.alterReplicaLogDirs(), the initial offset for future replica is also set to its HW. We update future replica's HW In ReplicaAlterLogDirsThread.processPartitionData(). We probably want to bound it by future replica's log end offset.

apovzner replied:
I fixed the comment to say it is either high watermark for future replica, or current replica's truncation offset. 
Also, changed ReplicaAlterLogDirsThread.processPartitionData() to bound future replica's high watermark to its log end offset.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,88 @@ abstract class AbstractFetcherThread(name: String,
+      warn(s"Based on $followerName's leader epoch, leader replied with an unknown offset in ${replica.topicPartition}. " +
+           s"The initial fetch offset ${partitionStates.stateValue(tp).fetchOffset} will be used for truncation.")
+      OffsetTruncationState(partitionStates.stateValue(tp).fetchOffset, truncationCompleted = true)
+    } else if (leaderEpochOffset.leaderEpoch == UNDEFINED_EPOCH || !useLeaderEpochInResponse) {
```
Do we need to test useLeaderEpochInResponse? The only case we want to cover here is that the follower uses version 0 of OffsetForLeaderEpoch.

apovzner replied:
I am testing it in `ReplicaFetcherThreadTest.shouldUseLeaderEndOffsetIfInterBrokerVersionBelow20`

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,88 @@ abstract class AbstractFetcherThread(name: String,
+      // less than or equal to the requested epoch.
+      val (replicaLeaderEpoch, replicaEndOffset) = replica.epochs.get.endOffsetFor(leaderEpochOffset.leaderEpoch)
+      if (replicaEndOffset == UNDEFINED_EPOCH_OFFSET) {
+        // This can happen if replica was not tracking offsets at that point (before the
```
not tracking offsets => not tracking leader epochs ?

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,88 @@ abstract class AbstractFetcherThread(name: String,
+        // we truncate to the end offset of the largest epoch that is smaller than the
+        // epoch the leader replied with, and send another offset for leader epoch request
+        val intermediateOffsetToTruncateTo = min(replicaEndOffset, replica.logEndOffset.messageOffset)
+        warn(s"Based on $followerName's leader epoch, leader replied with epoch ${leaderEpochOffset.leaderEpoch} " +
```
This is a normal behavior. So the logging probably should be info.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,88 @@ abstract class AbstractFetcherThread(name: String,
+      // its truncation offset)
+      warn(s"Based on $followerName's leader epoch, leader replied with an unknown offset in ${replica.topicPartition}. " +
+           s"The initial fetch offset ${partitionStates.stateValue(tp).fetchOffset} will be used for truncation.")
+      OffsetTruncationState(partitionStates.stateValue(tp).fetchOffset, truncationCompleted = true)
```
Now that we can truncate in more than 1 step, it's probably useful to always bound the truncation point by the replica's log end offset.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,88 @@ abstract class AbstractFetcherThread(name: String,
+        if (offsetToTruncateTo >= replica.logEndOffset.messageOffset) {
+          val logEndOffset = replica.logEndOffset.messageOffset
+          info(
+            s"Based on $followerName's leader epoch, leader replied with an offset $offsetFromLeader. Min(leader offset, leader epoch end offset on the follower) $offsetToTruncateTo >= the " +
```
Not sure if this logging is useful given the logging we have in log.truncateTo().

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -286,37 +287,41 @@ class ReplicaFetcherThread(name: String,
+   *  end offset of the largest epoch that is smaller than the epoch the leader replied with, and
+   *  send OffsetsForLeaderEpochRequest with that leader epoch. In a more rare case, where the
+   *  follower was not tracking epochs smaller than the epoch the leader replied with, we
+   *  truncate the leader's offset (and do not send any more leader epoch requesrs).
```
typo requesrs

junrao commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -164,17 +165,33 @@ class ReplicaAlterLogDirsThread(name: String,
+   *
+   * The logic for finding the truncation offset is the same as in ReplicaFetcherThread
+   * and mainly implemented in AbstractFetcherThread.getOffsetTruncationState. One difference is
+   * that the initial fetch offset for topic partition is truncation offset of the current replica.
```
Hmm, the initial fetch offset for future replica is set to its HW. Only when the current replica truncates, is the initial offset for future replica set to truncation offset of the current replica.

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -647,15 +647,17 @@ class Partition(val topic: String,
     * @param leaderEpoch Requested leader epoch
-    * @return The last offset of messages published under this leader epoch.
+    * @return The last offset of messages published under this leader epoch, or if the requested
+    *          leader epoch is unknown, the last offset of the largest epoch less than requested epoch
```
We want to mention that we are returning both the epoch and the offset.

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -28,6 +28,7 @@ import org.apache.kafka.common.metrics.Metrics
 import org.apache.kafka.common.protocol.Errors
 import org.apache.kafka.common.protocol.Errors._
 import org.apache.kafka.common.requests.EpochEndOffset
+import org.apache.kafka.common.requests.EpochEndOffset._
```
This exists in line 23 already.

apovzner replied:
removed dup from like 23

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -174,34 +176,216 @@ class ReplicaFetcherThreadTest {
+    val offsets = Map(t1p0 -> new EpochEndOffset(4, 155), t1p1 -> new EpochEndOffset(4, 143)).asJava
+
+    // Create the fetcher thread
+    val mockNetwork = new ReplicaFetcherMockBlockingSend(offsets, brokerEndPoint, new SystemTime())
```
This test should actually fail right now since the version of the leaderEpochRequest is alway version 1. So, we probably want to check the latestAllowedVersion() in the builder in ReplicaFetcherMockBlockingSend.

apovzner replied:
Right, this test ended up testing local broker on 0.11 and remote broker on latest version, which actually does not fail because we don't check leader epoch in leaderEpochResponse, and the truncation is done using KIP-101 approach (which is verified by this test). I will update the test to use undefined leader epoch in response to simulate another broker also on the older protocol version.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,88 @@ abstract class AbstractFetcherThread(name: String,
+    } else {
+      // get (leader epoch, end offset) pair that corresponds to the largest leader epoch
+      // less than or equal to the requested epoch.
+      val (replicaLeaderEpoch, replicaEndOffset) = replica.epochs.get.endOffsetFor(leaderEpochOffset.leaderEpoch)
```
replicaLeaderEpoch and leaderEpochOffset may be confusing. How about followerEpoch and leaderEpochOffset?

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -286,37 +287,41 @@ class ReplicaFetcherThread(name: String,
-    val fetchOffsets = scala.collection.mutable.HashMap.empty[TopicPartition, Long]
+   * Truncate the log for each partition's epoch based on leader's returned epoch and offset.
+   *  -- If the leader replied with undefined epoch offset, we must use the high watermark. This can
+   *  happen if 1) the leader is still using message format older than KAFKA_0_11_0_IV2; 2) the follower
```
KAFKA_0_11_0_IV2 => KAFKA_0_11_0

junrao reviewed (commented):
@apovzner : Thanks for the updated patch. A few more comments below. A couple of other things.

1. Could you also run the system tests?
2. This is not an issue directly related to this patch. But I noticed that in Log.truncateTo(), if the truncation point is in the middle of a message set, we will actually be truncating to the first offset of the message set. In that case, the replica fetcher thread should adjust the fetch offset to the actual truncated offset. Typically, the truncation point should never be in the middle of a message set. However, this could potentially happen during message format upgrade. We can tighten this up in a separate jira.

apovzner committed:
removed dup test due to merge, small cleanup to address couple of comments

apovzner committed:
Bounding future replica HW to its LOE, and some more cleanup to address comments

apovzner committed:
fixes to comments and log

apovzner committed:
updated unit test to properly simulate both brokers on older protocol version

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequest.java:
```diff
@@ -50,8 +50,11 @@
     private static final Schema OFFSET_FOR_LEADER_EPOCH_REQUEST_V0 = new Schema(
             new Field(TOPICS_KEY_NAME, new ArrayOf(OFFSET_FOR_LEADER_EPOCH_REQUEST_TOPIC_V0), "An array of topics to get epochs for"));
 
+    /* v2 request is the same as v1. Per-partition leader epoch has been added to response */
```
typo. Probably should be v1 instead of v2.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochResponse.java:
```diff
@@ -52,8 +54,23 @@
+    // OFFSET_FOR_LEADER_EPOCH_RESPONSE_PARTITION_V1 added a per-partition leader epoch field,
+    // which specifies which leader epoch the end offset belongs to
+    private static final Schema OFFSET_FOR_LEADER_EPOCH_RESPONSE_PARTITION_V1 = new Schema(
+        ERROR_CODE,
```
nits: can we make the indention same as the existing indention in this file?

lindong28 commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -77,6 +77,7 @@ object ApiVersion {
     // and KafkaStorageException for fetch requests.
     "1.1-IV0" -> KAFKA_1_1_IV0,
     "1.1" -> KAFKA_1_1_IV0,
+    // Introduced OffsetsForLeaderEpochRequest/OffsetsForLeaderEpochResponse V1 via KIP-279
```
nits: to be more consistent with the existing comment, we can just say `Introduced OffsetsForLeaderEpochRequest V1 via KIP-279`

lindong28 commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -647,15 +647,18 @@ class Partition(val topic: String,
   /**
     * @param leaderEpoch Requested leader epoch
-    * @return The last offset of messages published under this leader epoch.
+    * @return The requested leader epoch and the last offset of messages published under this
```
It looks like the existing Java doc (prior to this patch) of this method is not correct. According to Java doc of LeaderEpochFileCache.endOffsetFor(...), it says `The End Offset is the end offset of this epoch, which is defined as the start offset of the first Leader Epoch larger than the Leader Epoch requested, or else the Log End Offset if the latest epoch was requested`.

apovzner replied:
Yes, I think the prior description is more of a shortcut, which is actually not correct. I just realized that we should use "end offset" instead of the last offset of messages published here -- the description in LeaderEpochFileCache is more precise. I will update this comment accordingly.

lindong28 commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,84 @@ abstract class AbstractFetcherThread(name: String,
+    val followerName = if (isFutureReplica) "future replica" else "follower"
+
+    // Called when 'offsetToTruncateTo' is the final offset to truncate to.
+    def finalFetchLeaderEpochOffset(offsetToTruncateTo: Long, offsetFromLeader: Long): OffsetTruncationState = {
```
nits: now that we don't have any logging in finalFetchLeaderEpochOffset(), we can probably remove this method and replace its usage with one line. For example, `finalFetchLeaderEpochOffset(leaderEpochOffset.endOffset, leaderEpochOffset.endOffset)` is equivalent to `OffsetTruncationState(math.min(offsetToTruncateTo, replica.logEndOffset.messageOffset), truncationCompleted = true)`

lindong28 commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -54,7 +54,8 @@ class ReplicaFetcherThread(name: String,
                                 isInterruptible = false,
-                                includeLogTruncation = true) {
+                                includeLogTruncation = true,
+                                useLeaderEpochInResponse = brokerConfig.interBrokerProtocolVersion >= KAFKA_2_0_IV0) {
```
In ReplicaFetcherThread.fetchEpochsFromLeader(), the version of OffsetsForLeaderEpochRequest should probably be determined based on the interBrokerProtocolVersion. We can use OffsetsForLeaderEpochRequest V1 only if the interBrokerProtocolVersion >= KAFKA_2_0_IV0. Otherwise, we can rolling bounce the cluster to upgrade the code, the leader may still be running the old code and not recognizes OffsetsForLeaderEpochRequest V1.

apovzner replied:
The OffsetsForLeaderEpoch request is exactly the same in v0, so we don't need to explicitly check the protocol version when building the requests. If the leader is on older version, it will send v0 response, which will not include leader epoch field, which is handled in OffsetsForLeaderEpochResponse constructor by setting leader epoch field to undefined. In the fetcher thread, we handle this case (where the leader epoch is undefined) in maybeTruncate() and fall back to KIP-101 behavior, same as when this broker is on older protocol version.

lindong28 replied:
Hmm.. my understanding is that the version of the response should always match the version of the request. Thus in order to receive OffsetsForLeaderEpochResponse V1, the broker needs to send OffsetsForLeaderEpochRequest V1. And the broker should reject the request if the version of the request is not recognized. Did I miss something?

apovzner replied:
Yes, correct. I meant nothing different to do in the fetcher thread. If I understood the code correctly, ReplicaFetcherThread.fetchEpochsFromLeader() passes the OffsetsForLeaderEpochRequest.Builder to sendRequest(), and then build() is called on that builder with with a version in NetworkClient.doSend. It looks like the proper version will be used in that case.

lindong28 replied:
Currently if we do not explicitly specify the version for AbstractRequest.Builder(), the latest version of this request, as determined by ApiKeys.latestVersion(), will be used for this request. The latest version for OffsetsForLeaderEpochRequest will be V1 after this patch. we probably need to explicitly pass the version (determined by the IBP) to OffsetsForLeaderEpochRequest.Builder, similar to what we do for UpdateMetadataRequest.Builder() in ControllerBrokerRequestBatch.sendRequestsToBrokers().

apovzner replied:
Oh I see, thank you, let me take a look.

apovzner replied:
Thanks a lot for your help, I updated the code to use the OffsetsForLeaderEpochRequest version when building a request.

lindong28 commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,84 @@ abstract class AbstractFetcherThread(name: String,
+   */
+  def getOffsetTruncationState(tp: TopicPartition, leaderEpochOffset: EpochEndOffset, replica: Replica, isFutureReplica: Boolean = false): OffsetTruncationState = {
+    // to make sure we can distinguish log output for fetching from remote leader or local replica
+    val followerName = if (isFutureReplica) "future replica" else "follower"
```
nits: this replica can be either follower or future replica. Maybe the variable can be named `replicaName`?

apovzner replied:
Yeah, I already went back and forth couple of times regarding "replica" vs. "follower" (also re: your comment below). Jun commented (in this PR) that replica is also confusing in a way that leader is also a replica. And in case of future replica, it is also a follower, but of a different type. I propose to keep this name as is, but replace `replicaEndOffset` with `followerEndOffset` re: your comment below.

lindong28 commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,84 @@ abstract class AbstractFetcherThread(name: String,
+    } else {
+      // get (leader epoch, end offset) pair that corresponds to the largest leader epoch
+      // less than or equal to the requested epoch.
+      val (followerEpoch, replicaEndOffset) = replica.epochs.get.endOffsetFor(leaderEpochOffset.leaderEpoch)
```
nits: would the name `replicaEpoch` be more consistent with the name `replicaEndOffset`?

lindong28 commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -446,3 +516,10 @@ case class PartitionFetchState(fetchOffset: Long, delay: DelayedItem, truncating
+
+case class OffsetTruncationState(offset: Long, truncationCompleted: Boolean) {
+
+  def this (offset: Long) = this(offset, true)
```
nits: can we remove the space after `this`?

lindong28 commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -47,7 +50,8 @@ abstract class AbstractFetcherThread(name: String,
                                      isInterruptible: Boolean = true,
-                                     includeLogTruncation: Boolean)
+                                     includeLogTruncation: Boolean,
+                                     useLeaderEpochInResponse: Boolean = true)
```
My personal opinion is that it may be more general to just pass the `interBrokerProtocolVersion` to the constructor of `AbstactFetcherThread`. And we use this variable to determine the version of OffsetsForLeaderEpochRequest when we actually generate the builder for OffsetsForLeaderEpochRequest. It is more consistent with the existing usage `KafkaConfig.interBrokerProtocolVersion` in the code base. And if in the future there is some other logic that relies on the interBrokerProtocol in the AbstractFetcherThread, we won't need to add more variable to the constructor.

apovzner replied:
I agree. However, I just tried it, and it requires changes to ConsumerFetcherThread constructor, and then ConsumerFetcherManager, and so on. I think it would be easy to change later when we need more logic dependent on inter broker protocol version, and especially once we remove old consumer fetcher code.

lindong28 reviewed (commented):
Thanks for the update! Left a few comments

apovzner committed:
Various small fixes addressing review comments

apovzner committed:
Creating OffsetsForLeaderEpochRequests with version

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,74 @@ abstract class AbstractFetcherThread(name: String,
     * @param fetchOffsets the partitions to mark truncation complete
     */
-  private def markTruncationCompleteAndUpdateFetchOffset(fetchOffsets: Map[TopicPartition, Long]) {
+  private def markTruncationCompleteAndUpdateFetchOffset(fetchOffsets: Map[TopicPartition, OffsetTruncationState]) {
```
The method name now is not very accurate. It doesn't always mark truncation as completed.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,74 @@ abstract class AbstractFetcherThread(name: String,
+      warn(s"Based on $followerName's leader epoch, leader replied with an unknown offset in ${replica.topicPartition}. " +
+           s"The initial fetch offset ${partitionStates.stateValue(tp).fetchOffset} will be used for truncation.")
+      OffsetTruncationState(partitionStates.stateValue(tp).fetchOffset, truncationCompleted = true)
+    } else if (leaderEpochOffset.leaderEpoch == UNDEFINED_EPOCH || !useLeaderEpochInResponse) {
```
It seems that we don't really need the flag useLeaderEpochInResponse. If interBrokerProtocolVersion < KAFKA_2_0_IV0, it's guaranteed that leaderEpochOffset.leaderEpoch is UNDEFINED_EPOCH.

lindong28 replied:
@junrao I have thought about this as well. But for future replica, even if  interBrokerProtocolVersion < KAFKA_2_0_IV0, ReplicaAlterDirThread.fetchEpochsFromLeader() may still return EpochEndOffset whose leaderEpoch is not UNDEFINED_EPOCH. Maybe this method should also return UNDEFINED_EPOCH if interBrokerProtocolVersion < KAFKA_2_0_IV0?

apovzner replied:
Oh, thanks for raising this, Dong. I think we should then make ReplicaAlterDirThread.fetchEpochsFromLeader() return response with UNDEFINED_EPOCH to match "older protocol response".

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -286,37 +290,41 @@ class ReplicaFetcherThread(name: String,
+   *  truncate the leader's offset (and do not send any more leader epoch requests).
+   *  -- Otherwise, truncate to min(leader's offset, end offset on the follower for epoch that
+   *  leader replied with, follower's Log End Offset).
+   */
```
It seems that this comment is really for AbstractFetcherThread.getOffsetTruncationState(). If we move the comment there, we can also simplify the comment in ReplicaAlterLogDirsThread.maybeTruncate().

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -276,18 +280,74 @@ abstract class AbstractFetcherThread(name: String,
+      // less than or equal to the requested epoch.
+      val (followerEpoch, followerEndOffset) = replica.epochs.get.endOffsetFor(leaderEpochOffset.leaderEpoch)
+      if (followerEndOffset == UNDEFINED_EPOCH_OFFSET) {
+        // This can happen if replica was not tracking leader epochs at that point (before the
```
Since the code uses follower, perhaps we can say "if follower was not"

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -174,34 +175,217 @@ class ReplicaFetcherThreadTest {
     //We should have truncated to the offsets in the response
-    assertTrue(truncateToCapture.getValues.asScala.contains(156))
-    assertTrue(truncateToCapture.getValues.asScala.contains(172))
+    assertTrue("Expected offset 156 in captured truncation offsets " + truncateToCapture.getValues,
```
Perhaps we can change the text to sth like "Expect partition t1p0 to truncate to offset 156".

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -174,34 +175,217 @@ class ReplicaFetcherThreadTest {
+
+
+    //We should have truncated to the offsets in the first response
+    assertTrue("Expected offset 155 in captured truncation offsets " + truncateToCapture.getValues,
```
Should we further assert that the builder for OFFSET_FOR_LEADER_EPOCH in ReplicaFetcherMockBlockingSend.sendRequest() is set with the right version?

apovzner replied:
I modified ReplicaFetcherMockBlockingSend to save the version of OffsetsForLeaderEpochRequest and added couple of checks in the test.

junrao reviewed (commented):
@apovzner : Thanks for the latest patch. LGTM. Just a few more minor comments below.

apovzner committed:
Minor changes to address review comments

apovzner committed:
Future replica truncation is consistent with protocol version

junrao commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -164,17 +165,35 @@ class ReplicaAlterLogDirsThread(name: String,
-        tp -> new EpochEndOffset(Errors.NONE, replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch))
+        val (leaderEpoch, leaderOffset) = replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch)
+        val leaderEpochInResponse: Int =
+          if (brokerConfig.interBrokerProtocolVersion >= KAFKA_2_0_IV0) leaderEpoch
```
Do we need this check? Since we are getting the leader epoch from the current replica's log directly, even when IBP < KAFKA_2_0_IV0, it seems that we can just return leaderEpoch.

apovzner replied:
If we are on protocol < 2.0, then the local replica will be fetching from leader based on older protocol (not using leader epoch).  If we don't check here, the future replica will be fetching from the local replica based on leader epoch. Seems inconsistent? On the other hand, it should still work for the future replica to truncate using leader epoch in that  case too.

junrao replied:
Yes, preserving the leader epoch always gives better outcome. So, if we can do it, there is no reason to switch to a worse method. We have no choice btw follower and leader because of IBP. However, here, everything is local. So, there is no need to be constraint by IBP.

lindong28 replied:
If we keep the leader epoch here for better outcome, should we still check `useLeaderEpochInResponse` in getOffsetTruncationState() so that it returns `OffsetTruncationState(min(leaderEpochOffset.endOffset, replica.logEndOffset.messageOffset), truncationCompleted = true)` if `useLeaderEpochInResponse` is false?

apovzner replied:
If we use leader epoch, then we should go all the way using the new protocol, i.e., continue truncating until finding the consistent point. 
Ok, I will change back to using leader epoch if available for future replica.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -272,22 +275,96 @@ abstract class AbstractFetcherThread(name: String,
+        OffsetTruncationState(intermediateOffsetToTruncateTo, truncationCompleted = false)
+      } else {
+        val offsetToTruncateTo = min(followerEndOffset, leaderEpochOffset.endOffset)
+        OffsetTruncationState(min(offsetToTruncateTo, replica.logEndOffset.messageOffset), truncationCompleted = true)
```
In general, we don't expect the truncation point to be < local HW. So, it would be useful to log a warning when this happens. Not sure what's the easiest way since now we can have intermediate truncation point.

junrao reviewed (commented):
@apovzner : Thanks for the new patch. A couple of more comments.

apovzner committed:
future replica always uses leader epoch for truncation if available

apovzner committed:
reverted incorrect warn about truncating below HW

junrao commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -21,6 +21,7 @@ import java.nio.ByteBuffer
 import java.util
 
 import AbstractFetcherThread.ResultWithPartitions
+import kafka.api._
```
unused import

junrao reviewed (commented):
@apovzner : Thanks for the update. A minor comment below. Also, (1) have you added the logic to warn if truncation point < local HW? (2) have you run all systems?

apovzner committed:
Output warning if truncating below HW

apovzner commented:
@junrao I added the warning about truncating below HW to ReplicaFetcherThread.maybeTruncate. I explicitly compare replica.highWatermark to the offset we are truncating. If we truncate several times, and more than once below HW, we will output the warning multiple times, which I think is ok.

I ran system tests yesterday (https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1746/) and there was only one failure in kafkatest.benchmarks.streams.streams_simple_benchmark_test.StreamsSimpleBenchmarkTest.test_simple_benchmark.test=streams-join.scale=1 which was due to stream test process took too long to exit. I don't think it is related to any changes in this PR.

junrao reviewed (approved):
@apovzner : Thanks for the patch. LGTM.

guozhangwang commented on docs/upgrade.html:
```diff
@@ -19,6 +19,58 @@
 
 <script id="upgrade-template" type="text/x-handlebars-template">
 
+<h4><a id="upgrade_2_0_0" href="#upgrade_2_0_0">Upgrading from 0.8.x, 0.9.x, 0.10.0.x, 0.10.1.x, 0.10.2.x, 0.11.0.x, 1.0.x, 1.1.x, or 1.2.x to 2.0.0</a></h4>
```
@apovzner @junrao while working on another PR, I realized this one duplicated the part of `upgrade_2_0_0` with `upgrade_1_2_0` (we renamed 1.2 to 2.0). If there is nothing new content added I'll go ahead and remove the duplicated section in my PR
